src/storage/buffer_pool_manager/buffer_pool.rs

```diff
@@ -0,0 +1,432 @@
+            let page_table_item_read_lock_data_ref =
+                unsafe { page_table_item_read_lock_data.as_ref().unwrap() };
+
+            // Safety: We have a read lock over `page_table_item_read_lock`.
+            let PageTableItem::Occupied { metadata, frame_lock } = page_table_item_read_lock_data_ref else { panic!("Not occupied!") };
+            let frame_read_lock_raw = unsafe { frame_lock.raw() };
+            frame_read_lock_raw.lock_shared();
+
+            // Safety: We have a read lock over `frame_read_lock`.
+            let buffer = unsafe { self.frame_buffer.buffer.get().as_ref().unwrap() };
+            let data = buffer
+                .get(self.frame_buffer.get_frame_range(*buffer_id, 0))
+                .unwrap();
+
+            {
+                let mut page_table_item_metadata_lock = metadata.lock();
+                page_table_item_metadata_lock.access_count += 1;
+            }
+
+            PageReadLock::new(page_table_item_read_lock_raw, frame_read_lock_raw, data)
+        } else {
+            // TODO: Not in the buffer pool.
+            todo!()
+        }
+    }
+
+    fn mutate_page(&self, id: BufferPoolKey, offset: usize, data: Vec<u8>) {
+        assert!(offset % self.frame_buffer.page_size == 0);
+        assert!(offset + data.len() <= self.frame_buffer.page_size);
+
+        let buffer_id = self.buffer_map.get(&id).unwrap();
+
+        // Safety: `page_table` will never reallocate over the lifetime of the program.
+        let page_table = unsafe { self.page_table.table.get().as_ref().unwrap() };
+        let page_table_item_lock = page_table.get(*buffer_id).unwrap();
+
+        let page_table_item_read_lock = page_table_item_lock.read();
+
+        let PageTableItem::Occupied { ref metadata, ref frame_lock } = *page_table_item_read_lock else { panic!("Not occupied!") };
+        let frame_write_lock = frame_lock.write();
+
+        // Safety: We have a write lock over `frame_write_lock`.
+        let buffer = unsafe { self.frame_buffer.buffer.get().as_mut().unwrap() };
+        buffer
+            .get_mut(self.frame_buffer.get_frame_range(*buffer_id, offset))
+            .unwrap()
+            .copy_from_slice(&data);
+
+        {
+            let mut page_table_item_metadata_lock = metadata.lock();
+            page_table_item_metadata_lock.access_count += 1;
+            page_table_item_metadata_lock.flags.set_is_dirty(true);
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::{
+        sync::Arc,
+        thread::{sleep, spawn, JoinHandle},
+        time::Duration,
+    };
+
+    fn print_buffer_pool(buffer_pool: &BufferPoolImpl) {
+        println!("PageTable {:#?}", unsafe {
+            buffer_pool.page_table.table.get().as_ref().unwrap()
+        });
+        println!("FrameBuffer {:?}", unsafe {
+            buffer_pool.frame_buffer.buffer.get().as_ref().unwrap()
+        });
+        println!("{:#?}", buffer_pool);
+    }
+
+    #[test]
+    fn test() {
```

Comment:
Separate into multiple unit tests

Reply:
The tests depend on each other for locking/unlocking at different times, it might not be possible to separate them all. Also, this is a good semi-integration test to see if everything is working together.